drivers/charger/charger_bq25180.c

```diff
@@ -0,0 +1,210 @@
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t value = enable ? 0 : BQ25180_ICHG_CHG_DIS;
+	int ret;
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL,
+				     BQ25180_ICHG_CHG_DIS, value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int bq25180_set_charge_current(const struct device *dev,
+				      uint32_t const_charge_current_ua)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = bq25180_ma_to_ichg(const_charge_current_ua / 1000, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL,
+				     BQ25180_ICHG_MSK, val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int bq25180_get_charge_current(const struct device *dev,
+				      uint32_t *const_charge_current_ua)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*const_charge_current_ua = bq25180_ichg_to_ma(val) * 1000;
+
+	return 0;
+}
+
+static int bq25180_get_prop(const struct device *dev, charger_prop_t prop,
+			    union charger_propval *val)
+{
+	switch (prop) {
+	case CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA:
+		return bq25180_get_charge_current(dev, &val->const_charge_current_ua);
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int bq25180_set_prop(const struct device *dev, charger_prop_t prop,
+			    const union charger_propval *val)
+{
+	switch (prop) {
+	case CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA:
+		return bq25180_set_charge_current(dev, val->const_charge_current_ua);
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static const struct charger_driver_api bq25180_api = {
+	.get_property = bq25180_get_prop,
+	.set_property = bq25180_set_prop,
+	.charge_enable = bq25183_charge_enable,
+};
+
+static int bq25180_init(const struct device *dev)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_MASK_ID, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val &= BQ25180_DEVICE_ID_MSK;
+	if (val != BQ25180_DEVICE_ID) {
+		LOG_ERR("Invalid device id: %02x", val);
+		return -EINVAL;
+	}
+
+	/* Disable the watchdog */
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_IC_CTRL,
+				     BQ25180_WATCHDOG_SEL_1_MSK,
+				     BQ25180_WATCHDOG_DISABLE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (cfg->initial_current_microamp > 0) {
+		ret = bq25180_ma_to_ichg(cfg->initial_current_microamp / 1000, &val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL,
+					     BQ25180_ICHG_MSK, val);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+#define CHARGER_BQ25180_INIT(inst)								\
+	static const struct bq25180_config bq25180_config_##inst = {				\
+		.i2c = I2C_DT_SPEC_INST_GET(inst),						\
+		.initial_current_microamp = DT_INST_PROP(inst, initial_current_microamp),	\
```

Comment:
It would be better to use `constant_charge_current_max_microamp` from `battery.yaml`

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/67172

Reply:
brilliant, missed that file somehow, also missed to include `i2c-device.yaml`, fixed both

@bbilas I'm not sure what you are trying to suggest with that PR but it looks good to me :-)

@rriveramcrus how about moving `battery.yaml` under `charger`? It looks a bit lonely in that `battery/` directory with nothing else

Reply:
@fabiobaltieri `battery.yaml` lives in its own directory so that it can be shared by `charger` and `fuel_gauge`. 
The intention in the future is to have a single DT node for the battery that a systems charger(s) and fuel gauge could pick up characteristics from. Let me know if this is a good direction to work towards.

Reply:
ok fair enough, I think having a dedicated node for the battery is only worth if there's a corresponding substantial code base behind it that allows code reuse from different drivers, if it's just for having a common set of properties to pick from then I would just leave the common yaml file and let every driver do their own things, no point adding an extra node

Reply:
@fabiobaltieri I thought you were trying to add a custom solution for setting the initial current microamp for this driver instead of doing that in some generic way.

Reply:
Thanks for weighing in on this. On the charger side it would save some code during initialization, but that's not a whole lot of savings at this point so it's not a priority at the moment.

Reply:
I didn't notice that this is a charger, not a regulator... Sorry for the noise.

Reply:
yeah no prob, this was a regulator, I migrated it over